### ISSUE TITLE
Runtime parameters passed to input_fn for tf.estimator 

### ISSUE THREAD

**Issue by SumNeuron** [tensorflow/tensorflow#29789]
<em>Please make sure that this is a feature request. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:feature_template</em>

**System information**
- TensorFlow version (you are using): 1.10+ < 2
- Are you willing to contribute it (Yes/No): yes

**Describe the feature and the current behavior/state.**
Currently, to my knowledge, if one uses `tf.estimator.TrainSpec` / `tf.estimator.EvalSpec` the argument "`input_fn`" takes a function for producing either an iterator or something that returns a tuple of `(features, labels)`. However there does not seem to be a way to add epoch based variation to the dataset. 

It would be nice if TF would  automatically pass runtime parameters to this function e.g. via

```python
def input_fn(files:list, params:dict, _runtime_params:dict=None):
    # access to epoch via _runtime_params["epoch"]
``` 

While the use of epoch specific inputs during training will introduce a bottle neck it can be sometimes necessary.

**Will this change the current api? How?**

It will not break existing code. It will simply allow the user to have epoch / step specific control over the input function (e.g. if they want to upsample an underrepresented class from the last 200 epochs)

**Who will benefit with this feature?**
everyone who uses estimators

**Any Other info.**

**Comment by karmel** [comment]
Estimator does not make it easy to have different input fns based on context. @guptapriya -- is there a way to achieve this using the new 2.0 APIs?

**Comment by SumNeuron** [comment]
@karmel  my current solution isn't so much of a super solution... but maybe could be wrangled into one.... 

# context
The goal of the estimator API was to clearly separate the pre-processing pipeline, the data input pipeline, the training, and then inference.

The current estimator class however really only works as a scheduler for input, training/eval and inference. This forces the user to ensure the data is prepossessed thoroughly prior to the input function being called, or at the very least not require it to be called more than once. 

Yet there are many use cases where were batch / epoch specific modifications could be beneficial (weighting the next epoch based off of sampling bias from previous epoch, batch specific softening of labels, etc). 



# putative solutions

## hot fix and maybe easiest
A "hot-fix" would be changing the default of behavior. Currently, if steps (or an early stopping hook) is not provided, the model will run-indefinitely. An addition `RunConfig` parameter `loop_on_end_of_iteration` (or a snazzier variable name)  could be used to tell the Estimator to essentially recall `self.train` (and hence the `input_fn`) when terminating after hitting the end of the original `input_fn`. Of course this would have to auto-add `warm_start_from` to the call, so that the estimator picks up where it left off.

## maybe more in line with the api style
Right now `input_fn` is expected to be a  function that return `(feature, label)` or a dataset thereof.  It could however also expect a function that returns the `input_fn`.

This would, however, still require an overhaul to the estimator source code, for as far as I can tell `input_fn` is only called once.


## notes
in either case above, auto-passing some basic info to the `input_fn` (e.g. global step, batch number, epoch, etc) would allow for some nice dynamic datasets.

notable cons of both of these include:
- bottleneck now part of training
- all data must be in memory at once


# my meh workaround
At the moment, provided one has the luxury of disk space, is simply to precompute _n_  datasets and then have the `input_fn` uses those files appropriately (basically stack and treat as one large dataset)

## notes

a pro to this is that if one is using tfrecords all the benefits of tf records are still there (albeit with the aforementioned inconsistencies in [making and parsing them](https://github.com/tensorflow/tensorflow/issues/23802))

a con is that if one is not using tfrecords to stream from the files then your efforts go to waste unless you have an equally abundant amount of ram


# adjacent
to my knowledge in tf 2.0 the keras API also adopts the same standard for the `input_fn` and also shares this pitfall. So maybe this feature helps bolster estimator apis for the many with  custom estimators that will be around for years to come as well as move tf.keras forward?

**Comment by rmothukuru** [comment]
@SumNeuron,
Sorry for the delayed response. In the **`Tensorflow Version 2.x`**, since we use [TF Keras](https://www.tensorflow.org/api_docs/python/tf/keras/) and [tf.data](https://www.tensorflow.org/guide/data) predominantly and don't use [Estimators](https://www.tensorflow.org/guide/estimator) much, can you please let us know if this Feature is still relevant? 

Thanks!
